### PR TITLE
RFC: Adapt setools to the new libsepol's filename transition representation

### PR DESCRIPTION
**Pull request by WOnder93**
We are considering to merge SELinux userspace patches [1] that would cause setools to fail to build & link against libsepol. @jwcart2 already started working on abstracting a proper public API for use by setools instead of relying on libsepol's internals, but it may not be ready soon enough. This PR aims to prepare the patches that will be needed to make setools work with the new libsepol.

The first patch is intended only for testing - it directs the Travis CI to fetch userspace from a branch that contains the pending libsepol patches; the second patch does the actual conversion; and the last patch adds a few more filename transition rules to one of the test policies to exercise some corner cases in the new representation.

This PR is just an initial draft - feel free to suggest changes or different approaches. The goal is to at least have the changes ready and in mergeable state before the userspace patches are merged.

[1] https://lore.kernel.org/selinux/20200719103506.865962-1-omosnace@redhat.com/T/

### CODE REVIEW

File: tests/policyrep/selinuxpolicy.conf [hunk 1]
```diff
@@ -1650,6 +1650,10 @@ type_transition type95 type94:infoflow type95 "name95";
 type_transition type96 type95:infoflow type96 "name96";
 type_transition type97 type96:infoflow type97 "name97";
 
+type_transition type2 type96:infoflow type4 "name97";
+type_transition type3 type96:infoflow type4 "name97";
+type_transition type4 type96:infoflow type97 "name97";
```
Comment (fishilico) [inline tests/policyrep/selinuxpolicy.conf:1655]: Why is `"name97"` repeated instead of being incremented, like the lines above?

**Comment by fishilico** [comment]
Hi, I am not familiar with setools' codebase, but while testing @WOnder93's patches on SELinux's userspace, a test failed with a segmentation fault, which I found quite frustrating. The precise error was:
```console
~/projects/selinux-userspace/python/audit2allow $ make test
test_audit2allow (__main__.Audit2allowTests)
Verify audit2allow works ... ok
test_audit2why (__main__.Audit2allowTests)
Verify audit2why works ... FAIL
test_sepolgen_ifgen (__main__.Audit2allowTests)
Verify sepolgen-ifgen works ... ok
test_xperms (__main__.Audit2allowTests)
Verify that xperms generation works ... ok

======================================================================
FAIL: test_audit2why (__main__.Audit2allowTests)
Verify audit2why works
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_audit2allow.py", line 56, in test_audit2why
    self.assertSuccess("audit2why", p.returncode, err)
  File "test_audit2allow.py", line 24, in assertSuccess
    self.assertTrue(status == 0,
AssertionError: False is not true : "audit2why should have succeeded for this test None

----------------------------------------------------------------------
Ran 4 tests in 8.596s

FAILED (failures=1)
make: *** [Makefile:28: test] Error 1
```

This test failed because of a segmentation fault in setools:
```console
~/projects/selinux-userspace/python/audit2allow $ gdb -q -ex r --args python audit2why -p test_dummy_policy -i test.log
Reading symbols from python...
(No debugging symbols found in python)
Starting program: /usr/bin/python audit2why -p test_dummy_policy -i test.log
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[Detaching after fork from child process 1235408]

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff6a4deec in ?? () from /usr/lib/python3.8/site-packages/setools/policyrep.cpython-38-x86_64-linux-gnu.so
=> 0x00007ffff6a4deec:	8b 16	mov    (%rsi),%edx
(gdb) bt
#0  0x00007ffff6a4deec in ?? () from /usr/lib/python3.8/site-packages/setools/policyrep.cpython-38-x86_64-linux-gnu.so
#1  0x00007ffff6a506f0 in ?? () from /usr/lib/python3.8/site-packages/setools/policyrep.cpython-38-x86_64-linux-gnu.so
#2  0x00007ffff6a3cdb4 in ?? () from /usr/lib/python3.8/site-packages/setools/policyrep.cpython-38-x86_64-linux-gnu.so
#3  0x00007ffff7d07ecf in _PyObject_MakeTpCall () from /usr/lib/libpython3.8.so.1.0
#4  0x00007ffff7d03a78 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.8.so.1.0
#5  0x00007ffff7d0f7e8 in _PyFunction_Vectorcall () from /usr/lib/libpython3.8.so.1.0
#6  0x00007ffff7cfed08 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.8.so.1.0
#7  0x00007ffff7cfd654 in _PyEval_EvalCodeWithName () from /usr/lib/libpython3.8.so.1.0
#8  0x00007ffff7dadb03 in PyEval_EvalCode () from /usr/lib/libpython3.8.so.1.0
...
```

As far as I understand, this issue is addressed in this Pull Request. But this patch would make setools incompatible with the version of `libsepol` (last release, 3.1) installed on my development host.

Is it possible to support both libsepol 3.1 and libsepol + patch at <https://lore.kernel.org/selinux/20200719103506.865962-1-omosnace@redhat.com/T/>, in setools, at least during a transition period until libsepol 3.2 is released?

File: tests/policyrep/selinuxpolicy.conf (hunk 1 above)
Comment (WOnder93) [reply tests/policyrep/selinuxpolicy.conf:1655]: The intent is to have multiple rules with the same key under the new representation so that there are > 1 datums in that key's datum list (there is a separate datum entry for each `otype`) and also multiple `stype`s in at least one datum entry.

But your question made me realize that there should be a comment explaining that :)

**Comment by WOnder93** [comment]
> As far as I understand, this issue is addressed in this Pull Request. But this patch would make setools incompatible with the version of `libsepol` (last release, 3.1) installed on my development host.

Yes, that is almost certainly the reason.

> Is it possible to support both libsepol 3.1 and libsepol + patch at https://lore.kernel.org/selinux/20200719103506.865962-1-omosnace@redhat.com/T/, in setools, at least during a transition period until libsepol 3.2 is released?

If it were possible (without some complicated workaround), then I'd love to do it that way, but the fact that setools uses internal and unversioned structures and functions from libsepol makes that quite infeasible. I had suggested to link setools statically with libsepol, which would somewhat address the transition issue, but that wasn't met with acceptance:
https://lore.kernel.org/selinux/789178c6-3c52-c620-2d2f-ae2927a2fa86@ieee.org/

File: tests/policyrep/selinuxpolicy.conf (hunk 1 above)
Comment (pebenito) [reply tests/policyrep/selinuxpolicy.conf:1655]: I'm not clear what you're trying to test. This policy is meant to test that SETools is counting the number of rules correctly, by using prime number amounts of rules to make sure it's counting the right things. 100 doesn't exactly work as a prime :)

File: setools/policyrep/terule.pxi [hunk 2]
```diff
@@ -464,17 +464,18 @@ cdef class FileNameTERule(BaseTERule):
         readonly str filename
 
     @staticmethod
-    cdef inline FileNameTERule factory(SELinuxPolicy policy, sepol.filename_trans_t *key,
-                                       sepol.filename_trans_datum_t *datum):
+    cdef inline FileNameTERule factory(SELinuxPolicy policy,
+                                       sepol.filename_trans_key_t *key,
+                                       Type stype, size_t otype):
```
Comment (pebenito) [inline setools/policyrep/terule.pxi:469]: `stype` can only be a type and never an attribute?

**Comment by pebenito** [comment]
> Is it possible to support both libsepol 3.1 and libsepol + patch at https://lore.kernel.org/selinux/20200719103506.865962-1-omosnace@redhat.com/T/, in setools, at least during a transition period until libsepol 3.2 is released?

My initial reaction is to not merge this until much closer to libsepol 3.2 is released.

File: tests/policyrep/selinuxpolicy.conf (hunk 1 above)
Comment (WOnder93) [reply tests/policyrep/selinuxpolicy.conf:1655]: Hah, I didn't realize the numbers had some [special meaning](https://www.youtube.com/watch?v=QiVfqVQ5t9A) :) Well, I guess I could just remove three existing rules to make room for the new ones...

Anyway, I'm trying to test exactly that - that setools counts the number correctly even in the now "special" cases. The counting function gets more complex with the patch, so I'm just trying to keep the new corner cases covered.

File: setools/policyrep/terule.pxi (hunk 2 above)
Comment (WOnder93) [reply setools/policyrep/terule.pxi:469]: Yes, filename transitions are always expanded in the binary policy. I tried to add support for attributes there once, but I'd have to prove that it doesn't break performance in the kernel (looking up rules for types that belong to many attributes can be quite costly), which seemed like too much hassle at that time, so I gave up on that eventually.

**Comment by pebenito** [comment]
@WOnder93 since the 3.2 userspace release cycle is starting, can you update this?

**Comment by WOnder93** [comment]
Sorry, I had the follow-ups ready locally, just forgot to push them... The build passes with selinux userspace built from the master branch:
https://github.com/WOnder93/setools/actions/runs/465613941

**Comment by pebenito** [comment]
I think the plan is once 3.2 rc1 is out, you can update the build target in the PR tests to that userspace version so all these tests can pass. Please update the README to say it requires libsepol 3.2 (no "rc") too. Then this PR can merge cleanly.

**Comment by pebenito** [comment]
It seems ready to go. You should remove draft status if so.

**Comment by WOnder93** [comment]
Yes, I was just waiting for the CI to finish :)

**Comment by pebenito** [comment]
Thanks!
